src/Parsers/tests/gtest_format_hiliting.cpp

```diff
@@ -0,0 +1,303 @@
+#include <unordered_set>
+
+#include <Parsers/IAST.h>
+#include <Parsers/ParserQuery.h>
+#include <Parsers/parseQuery.h>
+#include <gtest/gtest.h>
+#include <Common/StackTrace.h>
+
+std::string hilite(const std::string & s, const char * hilite_type)
+{
+    std::stringstream ss;
+    ss << hilite_type;
+    ss << s;
+    ss << DB::IAST::hilite_none;
+    return ss.str();
+}
+
+std::string keyword(const std::string & s)
+{
+    return hilite(s, DB::IAST::hilite_keyword);
+}
+
+std::string identifier(const std::string & s)
+{
+    return hilite(s, DB::IAST::hilite_identifier);
+}
+
+std::string alias(const std::string & s)
+{
+    return hilite(s, DB::IAST::hilite_alias);
+}
+
+std::string op(const std::string & s)
+{
+    return hilite(s, DB::IAST::hilite_operator);
+}
+
+std::string function(const std::string & s)
+{
+    return hilite(s, DB::IAST::hilite_function);
+}
+
+std::string substitution(const std::string & s)
+{
+    return hilite(s, DB::IAST::hilite_substitution);
+}
+
+std::vector<const char *> HILITES =
+    {
+        DB::IAST::hilite_keyword,
+        DB::IAST::hilite_identifier,
+        DB::IAST::hilite_function,
+        DB::IAST::hilite_operator,
+        DB::IAST::hilite_alias,
+        DB::IAST::hilite_substitution,
+        DB::IAST::hilite_none
+    };
+
+[[maybe_unused]] const char * consume_hilites(const char ** it)
```

Comment:
I agree with the idea to keep the tests simple, but I'm not sure how to achieve the same results without the extra comparison logic.

There's no way to compare that the hilites match without writing a custom comparator.

E.g. one string has a space character hilited, the other one doesn't, and both are in accordance with the correct behaviour.

It would be infeasible to impose the invariant that e.g. whitespace shouldn't be hilited and each starting hilite should end with the ending hilite `hilite_none`. It would require me to refactor all formatImpl overrides, to ensure this invariant holds.

Also, I'm going to use this or similar code to regression-test my formatImpl refactoring #45649: I'm going to compare that the visible hiliting didn't change.

The most I can offer is to rewrite `are_equal_with_hilites_removed` since it's 1) unnecessary copies the string and 2) could possibly be implemented simpler. I'm just still not sure how to simplify it, thinking of throwing some `<algorithm>` function.

Reply:
I guess our notions of "hiliting equality" is different, I am in the bit-by-bit equality camp (to make developer life easy) whereas you lean to the "logical" equality camp. It is fine with me to go for logical equality.

Perhaps you could at least add a few code comments about the motivation for the "fuzzy" comparison, in case future readers wonder. And perhaps about the invariants.